src/Workspaces/Core/Portable/Shared/Utilities/StreamingProgressTracker.cs

```diff
@@ -0,0 +1,63 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Shared.Utilities
+{
+    /// <summary>
+    /// Utility class that can be used to track the progress of an operation in a threadsafe manner.
+    /// </summary>
+    internal class StreamingProgressTracker : IStreamingProgressTracker
+    {
+        private int _completedItems;
+        private int _totalItems;
+
+        private readonly Func<int, int, Task> _updateActionOpt;
+
+        public StreamingProgressTracker()
+            : this(null)
+        {
+        }
+
+        public StreamingProgressTracker(Func<int, int, Task> updateActionOpt)
+        {
+            _updateActionOpt = updateActionOpt;
+        }
+
+        public int CompletedItems => _completedItems;
+
+        public int TotalItems => _totalItems;
+
+        public Task AddItemsAsync(int count)
+        {
+            Interlocked.Add(ref _totalItems, count);
+            return UpdateAsync();
+        }
+
+        public Task ItemCompletedAsync()
+        {
+            Interlocked.Increment(ref _completedItems);
+            return UpdateAsync();
+        }
+
+        public Task ClearAsync()
```

Comment:
We do use it in the Sync-form for the preview dialog we show for 'fix all' refactorings.  You get a progress that says "computing fixes", then you have progress for "applying fixes".  The progress state is cleared between the two so you can gauge the progress of both.  We can't have only one progress because we don't know how many fixes we'll need to apply.  So, if we didn't clear, you'd see the 'computing fixes' move all the way to the right, then the addition of the fix count would make it jump back a bit, and then move forward.  

Having it start back at the beginning is a lot less jarring than jumping back a bit then moving forward.